### ISSUE TITLE
A warning should be displayed if an error occured during caching

### ISSUE THREAD

**Issue by kit-ty-kate** [ocaml/dune#9275]
## Expected Behavior

```
$ export DUNE_CACHE=enabled
$ du -hs /home/opam/.cache/dune
36K    /home/opam/.cache/dune
$ dune build
[Warn] An error occured while attempting to cache the result of the build.
$ du -hs /home/opam/.cache/dune
36K    /home/opam/.cache/dune
```

## Actual Behavior

No warning is displayed and the error is pushed under the rug.
I lost several hours trying to figure out why the dune cache wasn’t being filled even when dune cache was enabled

Turns out I was using the dune 2.0 `DUNE_CACHE_DUPLICATION=copy` environment variable instead of the dune 3.0 `DUNE_CACHE_STORAGE_MODE=copy`

## Reproduction

```
FROM ocaml/opam
ENV DUNE_CACHE=enabled
ENV DUNE_CACHE_TRANSPORT=direct
# ENV DUNE_CACHE_STORAGE_MODE=copy
RUN --mount=type=cache,id=dune-cache,target=/home/opam/.cache/dune,uid=1000,gid=1000 export DUNE_CACHE_STORAGE_MODE=copy && du -hs /home/opam/.cache/dune && opam install -vy base && du -hs /home/opam/.cache/dune
```

1. Run this dockerfile: fails silently
2. Run this dockerfile again with `ENV DUNE_CACHE_STORAGE_MODE=copy` uncommented: succeeds

## Specifications

- Version of `dune` (output of `dune --version`): 3.11.1
- Version of `ocaml` (output of `ocamlc --version`): 5.1.0
- Operating system (distribution and version): Debian (under docker)

**Comment by rgrinberg** [comment]
I think this suggestion bubbled up before but @snowleopard didn't like it. IIRC, the argument was that caching is just an optimization that should be transparent on success and failure

**Comment by kit-ty-kate** [comment]
It’s maybe an optimization but it’s a deliberate and required explicitly optimization so users should be able to see when there is a problem

**Comment by snowleopard** [comment]
I can imagine warning when an old environment variable like `DUNE_CACHE_DUPLICATION` is used. Not sure it's worth the maintenance cost to be honest since this particular variable is pretty old (and when it got deleted, the cache was still experimental, i.e. providing no promises of support?).

**Comment by ejgallego** [comment]
@kit-ty-kate does `_build/log` register the problem with the cache?

**Comment by kit-ty-kate** [comment]
@ejgallego yes but in a CI context you don’t have acces to that file. Plus regular users are not aware of this file and wouldn’t think about looking into it if they don’t even know there is an error until they notice long after.
